scene/gui/tree.cpp

```diff
@@ -247,31 +247,25 @@ void TreeItem::_propagate_check_through_parents(int p_column, bool p_emit_signal
 		return;
 	}
 
-	bool all_unchecked_and_not_indeterminate = true;
-	bool any_unchecked_or_indeterminate = false;
+	bool has_checked = false;
+	bool has_unchecked = false;
+	bool has_indeterminate = false;
 
 	TreeItem *child_item = current->get_first_child();
-	while (child_item) {
-		if (!child_item->is_checked(p_column)) {
-			any_unchecked_or_indeterminate = true;
-			if (child_item->is_indeterminate(p_column)) {
-				all_unchecked_and_not_indeterminate = false;
-				break;
-			}
+	while(child_item) {
+		if (child_item->is_checked(p_column)) {
+			has_checked = true;
+		} else if (child_item->is_indeterminate(p_column)) {
+			has_indeterminate = true;
```

Comment:
```suggestion
			any_indeterminate = true;
			break;
```
Finding an indeterminate makes the check only end one way, so we can exit early